### PR TITLE
CORTX-27991: Update storage resources(CVG, Disk) health status in System Health

### PR DESCRIPTION
**Pull request by mariyappanp**
Update storage resources(CVG, Disk) health status in System Health

# Problem Statement
- Problem statement

# Design
https://jts.seagate.com/browse/CORTX-27991
LLD: https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/891781277/Action+Status+Update+LLD

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [x] Unit and System Tests are added
https://jts.seagate.com/secure/attachment/509955/cvg_test.txt
https://jts.seagate.com/secure/attachment/509956/disk_test.txt

- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained
- [ ] Is there a change in filename/package/module or signature? [Y/N]: 
- [ ] If yes for above point, is a notification sent to all other cortx components? [Y/N]
- [ ] Side effects on other features (deployment/upgrade)? [Y/N]
- [ ] Dependencies on other component(s)? [Y/N]
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: ha/core/system_health/status_mapper.py [hunk 1]
```diff
@@ -51,7 +52,8 @@ class StatusMapper:
         "repairing"   : "informational",
         "repaired"    : "informational",
         "rebalancing" : "informational",
-        "failed"      : "error"
+        "failed"      : "error",
+        "offline"     : "informational"
```
Comment (mukhtar-inamdar) [inline ha/core/system_health/status_mapper.py:56]: Is this need to add as it is provisional and may not be required in code?
@ArchanaLimaye  can you please confirm?

File: ha/core/system_health/system_health.py [hunk 2]
```diff
@@ -393,7 +394,7 @@ def process_event(self, healthevent: HealthEvent):
                                         cluster_id=healthevent.cluster_id, site_id=healthevent.site_id,
                                         rack_id=healthevent.rack_id, storageset_id=healthevent.storageset_id,
                                         node_id=healthevent.node_id, server_id=healthevent.node_id,
-                                        storage_id=healthevent.node_id)
+                                        storage_id=healthevent.node_id, cvg_id=self.cvg_id)
```
Comment (mukhtar-inamdar) [inline ha/core/system_health/system_health.py:397]: same as above

**Comment by akash2144** [comment]
Can you please rebase this branch with latest fault_k8s. (PR 667 was merged)

File: ha/core/system_health/status_mapper.py (hunk 1 above)
Comment (shaileshsvc) [reply ha/core/system_health/status_mapper.py:56]: Offline is required if disk is marked offline for investigation.

Comment (mukhtar-inamdar) [reply ha/core/system_health/status_mapper.py:56]: ok

File: ha/core/system_health/system_health.py [hunk 3]
```diff
@@ -318,7 +318,7 @@ def _update(self, healthevent: HealthEvent, healthvalue: str, next_component: st
         key = self._prepare_key(component, cluster_id=self.node_map['cluster_id'], site_id=self.node_map['site_id'],
                                 rack_id=self.node_map['rack_id'], storageset_id=self.node_map['storageset_id'],
                                 node_id=self.node_id, server_id=self.node_id, storage_id=self.node_id,
-                                comp_type=comp_type, comp_id=comp_id)
+                                comp_type=comp_type, comp_id=comp_id, cvg_id=self.cvg_id)
```
[the remark this answers is not in the document]
Comment (mukhtar-inamdar) [reply ha/core/system_health/system_health.py:321]: i think has been checked-in already then leave it

File: ha/core/system_health/system_health.py (hunk 2 above)
Comment (mukhtar-inamdar) [reply ha/core/system_health/system_health.py:397]: I think this already checked-in so, you can leave it

File: ha/core/system_health/status_mapper.py [hunk 4]
```diff
@@ -63,7 +65,9 @@ def __init__(self):
     def map_event(self, event: HealthEvent) -> str:
         """Returns the status by mapping it against the source and event type."""
         try:
-            if event.source == HEALTH_EVENT_SOURCES.MONITOR.value and event.event_type == HEALTH_EVENTS.ONLINE.value:
+            component_type = event.resource_type.split(':')[-1]
+            if event.source == HEALTH_EVENT_SOURCES.MONITOR.value and event.event_type == HEALTH_EVENTS.ONLINE.value and \
+                component_type in [CLUSTER_ELEMENTS.CLUSTER.value, CLUSTER_ELEMENTS.NODE.value]:
```
Comment (Madhura-08) [inline ha/core/system_health/status_mapper.py:70]: Why this additional check is needed?

File: ha/k8s_setup/ha_setup.py [hunk 5]
```diff
@@ -264,6 +264,8 @@ def process(self):
 
             Log.info(f'Populating the ha config file with consul_endpoint: {consul_endpoint}')
 
+            Log.info('Performing event_manager subscription')
+            event_manager = EventManager.get_instance()
```
Comment (Madhura-08) [inline ha/k8s_setup/ha_setup.py:268]: Why it is showing as a new change? I see that its already there!

File: ha/core/system_health/system_health.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mariyappanp) [reply ha/core/system_health/system_health.py]: Rebased.

File: ha/core/system_health/system_health.py (hunk 2 above)
Comment (mariyappanp) [reply ha/core/system_health/system_health.py]: Rebased.

File: ha/k8s_setup/ha_setup.py (hunk 5 above)
Comment (mariyappanp) [reply ha/k8s_setup/ha_setup.py]: Rebased.

File: ha/core/system_health/status_mapper.py (hunk 4 above)
Comment (Madhura-08) [reply ha/core/system_health/status_mapper.py]: Ok..But why cluster is also included in the check?

File: ha/k8s_setup/ha_setup.py (hunk 5 above)
Comment (Madhura-08) [reply ha/k8s_setup/ha_setup.py]: Ok..

File: ha/core/system_health/status_mapper.py (hunk 4 above)
Comment (Madhura-08) [reply ha/core/system_health/status_mapper.py]: Oh Ok..You have removed that cluster as well..!!

**Review by shaileshsvc — APPROVED** [review]
Approved

**Review by Madhura-08 — COMMENTED** [review]
please add details for system health keys. Also, if you have any, please add test case for execution of "on_failure".

**Comment by ArchanaLimaye** [comment]
@mariyappanp please confirm that following is tested and update the respective outputs in test details(where applicable) 

- hierarchical updates are not happening for disk / CVG events 
- In case of CVG events, health update is happening and event is sent to health monitor 
 In case of  disk events, health update is happening and event is sent to health monitor 
- consul outputs show correct keys with correct status in all above scenarios

**Review by Madhura-08 — APPROVED** [review]

**Review by gauri-bhosale — APPROVED** [review]

**Review by akash2144 — APPROVED** [review]

**Review by mukhtar-inamdar — APPROVED** [review]

File: ha/core/health_monitor/health_monitord.py [hunk 6]
```diff
@@ -74,6 +75,8 @@ def __init__(self, singleton_check: bool = False):
         self._confstore = ConfigManager.get_confstore()
         self._rule_manager = MonitorRulesManager()
         self._event_consumer = self._get_consumer()
+        # Initialize conf store
+        ConftStoreSearch()
```
Comment (ArchanaLimaye) [inline ha/core/health_monitor/health_monitord.py:79]: Why is this needed in Health monitor ?

File: ha/util/conf_store.py [hunk 7]
```diff
@@ -221,3 +227,25 @@ def get_disk_list_for_cvg(index, node_id, cvg_id):
...
+
+        >>> ConftStoreSearch.get_cvg_for_disk('21d6291109304485b3daff43a06cff77', '/dev/sdd')
+        'cvg-01'
+        """
+        cvg_id = ""
+        cvg_list = ConftStoreSearch.get_cvg_list(CLUSTER_GLOBAL_INDEX, node_id)
+        for cvg_id in cvg_list:
+            if disk_id in ConftStoreSearch.get_disk_list_for_cvg(CLUSTER_GLOBAL_INDEX, node_id, cvg_id):
+                return cvg_id
+        return cvg_id
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:251]: cvg_id should not be returned here right  ? it should be returned only on line 250 I guess

File: ha/util/conf_store.py [hunk 8]
```diff
@@ -28,17 +28,23 @@
...
     """
     Wrapper class for using the confstore search APIs
     """
+    _cluster_conf_loaded = False
 
     def __init__(self, conf_store_req: bool = True):
         """Init method"""
         if conf_store_req:
             self._confstore = ConfigManager.get_confstore()
+        if not ConftStoreSearch._cluster_conf_loaded:
```
Comment (ArchanaLimaye) [inline ha/util/conf_store.py:45]: Please create the new ticket as discussed. In the DoD along with other changes, also add a point that the relevant code added in this PR should be removed.

File: ha/core/health_monitor/health_monitord.py (hunk 6 above)
Comment (mariyappanp) [reply ha/core/health_monitor/health_monitord.py]: Not required, but added for all services can avail this.
Removed.

File: ha/core/system_health/status_mapper.py [hunk 9]
```diff
@@ -63,7 +65,9 @@ def __init__(self):
     def map_event(self, event: HealthEvent) -> str:
         """Returns the status by mapping it against the source and event type."""
         try:
-            if event.source == HEALTH_EVENT_SOURCES.MONITOR.value and event.event_type == HEALTH_EVENTS.ONLINE.value:
+            component_type = event.resource_type.split(':')[-1]
```
[the remark this answers is not in the document]
Comment (mariyappanp) [reply ha/core/system_health/status_mapper.py]: Chnaged to `component_type = event.resource_type`

File: ha/util/conf_store.py (hunk 7 above)
Comment (mariyappanp) [reply ha/util/conf_store.py]: It would return empty string in case not entered into loop.
But fine to just mention only inside the loop.

File: ha/util/conf_store.py (hunk 8 above)
Comment (mariyappanp) [reply ha/util/conf_store.py:45]: sure

File: ha/util/conf_store.py (hunk 7 above)
Comment (ArchanaLimaye) [reply ha/util/conf_store.py]: Thanks. It would not return empty string if cvg_id is valid but disk id is not present inside this cvg_id and disk-id is not present in any of the cvg_ids. In this case incorrect cvg)id would be returned

File: ha/util/conf_store.py (hunk 8 above)
Comment (ArchanaLimaye) [reply ha/util/conf_store.py:45]: https://jts.seagate.com/browse/CORTX-29597 is created

**Review by ArchanaLimaye — APPROVED** [review]

**Review by gauri-bhosale — APPROVED** [review]

File: ha/util/conf_store.py (hunk 8 above)
Comment (akash2144) [inline ha/util/conf_store.py:45]: Why do we need this check ?

File: ha/util/conf_store.py [hunk 10]
```diff
@@ -221,3 +227,24 @@ def get_disk_list_for_cvg(index, node_id, cvg_id):
...
             Log.error(f"Unable to fetch Disk list from GConf. Error {e}")
             raise Exception(f"Unable to fetch Disk list. Error {e}")
+
+
+    @staticmethod
+    def get_cvg_for_disk(node_id, disk_id):
+        """
+        Returns cvg id for given node_id and disk_id.
+        Args:
+            index(str): index of conf file
```
Comment (akash2144) [inline ha/util/conf_store.py:237]: We don't need this index args here, right ?

**Review by akash2144 — CHANGES_REQUESTED** [review]

File: ha/util/conf_store.py (hunk 8 above)
Comment (mariyappanp) [reply ha/util/conf_store.py:45]: One time loading purpose.

Comment (mukhtar-inamdar) [reply ha/util/conf_store.py:45]: @mariyappanp, yes you don't need this check as the ConfigManager._safe_load() do this all

File: ha/util/conf_store.py (hunk 10 above)
Comment (mariyappanp) [reply ha/util/conf_store.py:237]: Yes, it has become an obsolete statement, I will remove it,

File: ha/util/conf_store.py (hunk 8 above)
Comment (mariyappanp) [reply ha/util/conf_store.py]: Done.

File: ha/util/conf_store.py (hunk 10 above)
Comment (mariyappanp) [reply ha/util/conf_store.py]: Done.

**Review by akash2144 — APPROVED** [review]

File: ha/fault_tolerance/fault_tolerance_driver.py [hunk 11]
```diff
@@ -40,6 +41,7 @@ def __init__(self):
         ConfigManager.init("fault_tolerance")
         self.node_fault_monitor = HealthStatusMonitor()
         self.cluster_stop_monitor = ClusterStopMonitor()
+        ConftStoreSearch()
```
Comment (Madhura-08) [inline ha/fault_tolerance/fault_tolerance_driver.py:44]: Discussed with Mariyappan for this! It is introduced here so that in system health, setach API call will be used to get cvg ids.
It is temporary code for now. Will be removed in next ticket.

File: ha/util/conf_store.py [hunk 12]
```diff
@@ -39,6 +41,8 @@ def __init__(self, conf_store_req: bool = True):
         """Init method"""
         if conf_store_req:
             self._confstore = ConfigManager.get_confstore()
+        # Load cluster config
+        ConfigManager._safe_load(CLUSTER_GLOBAL_INDEX, CLUSTER_CONF)
```
Comment (Madhura-08) [inline ha/util/conf_store.py:45]: Once. keys will be stored in consul which HA mini provisioning, this step will not be needed. Mariyappan confirmed.

File: ha/core/system_health/system_health.py [hunk 13]
```diff
@@ -385,7 +386,16 @@ def process_event(self, healthevent: HealthEvent):
 
             # Update the node map
             self.node_id = healthevent.node_id
-            self.cvg_id = healthevent.specific_info[NODE_MAP_ATTRIBUTES.CVG_ID.value] if component_type == CLUSTER_ELEMENTS.DISK.value else None
+            self.cvg_id = None
+            if component_type == CLUSTER_ELEMENTS.DISK.value:
+                if not isinstance(healthevent.specific_info, dict):
+                    healthevent.specific_info = {}
+                if not healthevent.specific_info.get(NODE_MAP_ATTRIBUTES.CVG_ID.value):
+                    self.cvg_id = ConftStoreSearch.get_cvg_for_disk(self.node_id, component_id)
```
Comment (Madhura-08) [inline ha/core/system_health/system_health.py:394]: Instead of using ConfSerarch API, keys will be fetched directly from consul..!

**Review by Madhura-08 — APPROVED** [review]

**Review by shaileshsvc — COMMENTED** [review]
Approved (Usage of cluster.conf vs consul to be taken up separately)
